### PR TITLE
Add support for custom-named helpers factories to avoid class names collision

### PR DESCRIPTION
**Pull request by almozavr**
- one can use `-Ajanet.http.factory.class.suffix=MyLib` for processor to generate factory class with suffix;
- helpers factory class is generate even if no @HttpAction classes are present (to import libs, e.g.);
- helpers classes are not generated if already present;

### CODE REVIEW

File: service-compiler/src/main/java/io/techery/janet/HttpHelpersGenerator.java [hunk 1]
```diff
@@ -57,12 +57,18 @@
...
-            while (actionClass != null
-                    && !checkSet.contains(actionClass.getTypeElement())) {
-                generate(actionClass);
-                checkSet.add(actionClass.getTypeElement());
+            while (actionClass != null && !processedElements.contains(actionClass.getTypeElement())) {
+                // generate only if there is no such helper class yet
+                try {
+                    String name = actionClass.getHelperName().packageName() +
+                            "." + actionClass.getHelperName().simpleName();
+                    Class.forName(name, false, getClass().getClassLoader());
```
Comment (dirong) [inline service-compiler/src/main/java/io/techery/janet/HttpHelpersGenerator.java]: Can we check class existing using `Elements.getTypeElement("....")`?  Maybe it will be better performance.

File: service-compiler/src/main/java/io/techery/janet/JanetHttpProcessor.java [hunk 2]
```diff
@@ -31,28 +34,31 @@
...
+        httpHelpersFactoryGenerator = new HttpHelpersFactoryGenerator(filer, findOtherHelpersFactories(), options);
         httpHelpersGenerator = new HttpHelpersGenerator(filer);
     }
 
     @Override
     public Set<String> getSupportedAnnotationTypes() {
-        Set<String> annotataions = new HashSet<String>();
-        annotataions.add(HttpAction.class.getCanonicalName());
-        return annotataions;
+        return Collections.singleton("*");
```
Comment (dirong) [inline service-compiler/src/main/java/io/techery/janet/JanetHttpProcessor.java:76]: Are you sure that we should process all annotations? 
Why did you change it?

File: service-compiler/src/main/java/io/techery/janet/JanetHttpProcessor.java [hunk 3]
```diff
@@ -75,11 +83,29 @@ public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment
...
+
+    private List<TypeElement> findOtherHelpersFactories() {
+        PackageElement packageElement = elementUtils.getPackageElement(HttpActionService.HELPERS_FACTORY_CLASS_PACKAGE);
+        if (packageElement == null) return Collections.emptyList();
+        //
+        List<TypeElement> factoryElements = new ArrayList<TypeElement>();
+        for (Element element : packageElement.getEnclosedElements()) {
+            if (element.getKind().isInterface()) continue;
+            for (TypeMirror typeMirror : ((TypeElement) element).getInterfaces()) {
+                TypeElement interfaceElement = elementUtils
```
Comment (dirong) [inline service-compiler/src/main/java/io/techery/janet/JanetHttpProcessor.java]: `typeUtils.asElement(typeMirror)`

File: service-compiler/src/main/java/io/techery/janet/JanetHttpProcessor.java (hunk 2 above)
Comment (almozavr) [inline service-compiler/src/main/java/io/techery/janet/JanetHttpProcessor.java:76]: Because, sometimes we have no `@HttpAction` annotated classes but have to search for other factories. Processor doesn't call `process` if there is no classes annotated with `@SupportedAnnotationTypes` values

I'm sorry didn't mentioned this with pr description.

File: service-compiler/src/main/java/io/techery/janet/HttpHelpersGenerator.java (hunk 1 above)
Comment (almozavr) [inline service-compiler/src/main/java/io/techery/janet/HttpHelpersGenerator.java]: yes, I think we can

Comment (almozavr) [inline service-compiler/src/main/java/io/techery/janet/HttpHelpersGenerator.java]: but that would mean we'll have to pass Elements instance to generator class. Do we want it?

File: service-compiler/src/main/java/io/techery/janet/HttpHelpersFactoryGenerator.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        if (shouldIncludeOtherFactories()) {
+            // add field for other factories and initialize it in constructor
+            ParameterizedTypeName listType = ParameterizedTypeName.get(ArrayList.class, HttpActionService.ActionHelperFactory.class);
+            classBuilder.addField(
+                    FieldSpec.builder(listType, "otherFactories", Modifier.PRIVATE)
+                            .initializer("new $T()", listType)
+                            .build()
+            );
+            for (TypeElement factoryElement : otherFactoriesElements) {
+                if (factoryElement.getQualifiedName().equals(className)) continue;
```
Comment (dirong) [inline service-compiler/src/main/java/io/techery/janet/HttpHelpersFactoryGenerator.java]: Comparing qualified name with simple name. Use method `getSimpleName()` to do it. But this method returns `Name` and you should convert it to `String` before comparing.

File: service-compiler/src/main/java/io/techery/janet/JanetHttpProcessor.java (hunk 3 above)
Comment (almozavr) [inline service-compiler/src/main/java/io/techery/janet/JanetHttpProcessor.java]: fuckn github, I don't understand what line are you want to change with this?

File: service-compiler/src/main/java/io/techery/janet/HttpHelpersFactoryGenerator.java (hunk 4 above)
Comment (almozavr) [inline service-compiler/src/main/java/io/techery/janet/HttpHelpersFactoryGenerator.java]: That was typo, sorry. I want to exclude canonical factory class by it's name. It could only happen if we generate general (no suffix) factory

File: service-compiler/src/main/java/io/techery/janet/HttpHelpersGenerator.java (hunk 1 above)
Comment (dirong) [inline service-compiler/src/main/java/io/techery/janet/HttpHelpersGenerator.java]: Sometimes  we want to get some information about some element or type. 
Maybe we should put this information into `HttpActionClass`. What do you think about it?

Comment (almozavr) [inline service-compiler/src/main/java/io/techery/janet/HttpHelpersGenerator.java]: do you mean smth like `HttpActionClass#exists()`?

File: service-compiler/src/main/java/io/techery/janet/JanetHttpProcessor.java (hunk 2 above)
Comment (dirong) [inline service-compiler/src/main/java/io/techery/janet/JanetHttpProcessor.java]: I see. We should generate a factory decorator for other factories from depended libraries. 
I think the better way will be adding some annotation to invoke processor for this operation.

File: service-compiler/src/main/java/io/techery/janet/JanetHttpProcessor.java (hunk 3 above)
Comment (dirong) [inline service-compiler/src/main/java/io/techery/janet/JanetHttpProcessor.java]: instead of that `elementUtils.getTypeElement(HttpActionService.ActionHelperFactory.class.getCanonicalName());`

Comment (almozavr) [inline service-compiler/src/main/java/io/techery/janet/JanetHttpProcessor.java]: still can't get a point

Comment (almozavr) [inline service-compiler/src/main/java/io/techery/janet/JanetHttpProcessor.java]: We have to get an Element from canonical class, not from a TypeMirror

**Comment by almozavr** [comment]
@dirong uploaded some changes we discussed in comments

**Comment by dirong** [comment]
LGTM
